protocols/direct-funding.go

```diff
@@ -0,0 +1,378 @@
+		if len(effects) > 0 {
+			// todo: effects as SideEffects{}
+			return updated, NoSideEffects, WaitingForCompleteFunding, nil
+		}
+	}
+
+	if !fundingComplete {
+		return updated, NoSideEffects, WaitingForCompleteFunding, nil
+	}
+
+	// Postfunding
+	if !updated.PostFundSigned[updated.MyIndex] {
+		// todo: []string{SignPostFundEffect(updated.ChannelId)} as SideEffects{}
+		return updated, NoSideEffects, WaitingForCompletePostFund, nil
+	}
+
+	if !updated.postfundComplete() {
+		return updated, NoSideEffects, WaitingForCompletePostFund, nil
+	}
+
+	// Completion
+	// todo: []string{"Objective" + s.ChannelId.String() + "complete"} as SideEffects{}
+	return updated, NoSideEffects, WaitingForNothing, nil
+}
+
+//  Private methods on the DirectFundingObjectiveState
+
+// prefundComplete returns true if all participants have signed a prefund state, as reflected by the extended state
+func (s DirectFundingObjectiveState) prefundComplete() bool {
+	for _, index := range s.ParticipantIndex {
+		if !s.PreFundSigned[index] {
+			return false
+		}
+	}
+	return true
+}
+
+// postfundComplete returns true if all participants have signed a postfund state, as reflected by the extended state
+func (s DirectFundingObjectiveState) postfundComplete() bool {
+	for _, index := range s.ParticipantIndex {
+		if !s.PostFundSigned[index] {
+			return false
+		}
+	}
+	return true
+}
+
+// fundingComplete returns true if the supplied onChainHoldings are greater than or equal to the threshold for being fully funded.
+func (s DirectFundingObjectiveState) fundingComplete(onChainHoldings types.Funds) bool {
+	for asset, threshold := range s.FullyFundedThreshold {
+		chainHolding, ok := onChainHoldings[asset]
+
+		if !ok {
+			return false
+		}
+
+		if gt(threshold, chainHolding) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// safeToDeposit returns true if the supplied onChainHoldings are greater than or equal to the threshold for safety.
+func (s DirectFundingObjectiveState) safeToDeposit(onChainHoldings types.Funds) bool {
+	for asset, safetyThreshold := range s.MyDepositSafetyThreshold {
+		chainHolding, ok := onChainHoldings[asset]
+
+		if !ok {
+			return false
+		}
+
+		if gt(safetyThreshold, chainHolding) {
+			return false
+		}
+	}
+
+	return true
+}
+
+// amountToDeposit computes the appropriate amount to deposit using the supplied onChainHolding
+func (s DirectFundingObjectiveState) amountToDeposit(onChainHoldings types.Funds) types.Funds {
+	deposits := make(types.Funds, len(onChainHoldings))
+
+	for asset, holding := range onChainHoldings {
+		deposits[asset] = big.NewInt(0).Sub(s.MyDepositTarget[asset], holding)
+	}
+
+	return deposits
+}
+
+// SignatureReceived updates the objective's cache of which participants have signed which states
+func (s DirectFundingObjectiveState) signatureRecieved(signature state.Signature, turnNum int) (DirectFundingObjectiveState, error) {
+	updated := s.clone()
```

Comment:
I think we don't need to clone, here (nor return a `DirectFundingObjectiveState`). All of the private methods can mutate the receiver, as long as the top level public method passes in a clone?

We can enforce that the public methods return a copy in a test.

Reply:
Allowing the private methods to make modifications something like https://play.golang.org/p/fPIw2wsAnMX ?

Reply:
Yes. IMO the important thing is that the part of the class that implements the `Objective` interface (i.e. the `Update` and the `Crank`) strictly adheres to the functional paradigm. It must return a copy of the receiver.

The rest is implementation detail -- it is "not wrong" to create unnecessary clones as the public method calls into various private methods. But it is also "not wrong" to allow the private methods to mutate the receiver if the receiver is already a clone generated in the top level public method -- because this wouldn't violate the rule above. It would also be slightly neater and more efficient, although I think these are secondary considerations just now. Does that make sense?

Reply:
Definitely agree that the `clone` can be dropped and enforced at the interface boundary. A little apprehensive about ending up with a mix of methods which mutate the caller and methods that don't.

How is 151825c?

Reply:
I share your apprehension. We can at least make sure the docstrings / comments stipulate whether the method mutates or not, to reduce the risk of confusion or a mistake. 151825c looks good to me.